modules/play-json/src/main/scala/PlayJsonModule.scala

```diff
@@ -0,0 +1,170 @@
+    }
+
+  final val labelRecordFields: HCoalgebra[HEnvT[Boolean, RSchema, ?[_], ?], LabelledSchema] =
+    new (LabelledSchema ~> HEnvT[Boolean, RSchema, LabelledSchema, ?]) {
+
+      def apply[A](seed: LabelledSchema[A]): HEnvT[Boolean, RSchema, LabelledSchema, A] =
+        seed match {
+          case (x, Fix(r @ Record(_, _))) => HEnvT(x, r.hmap(ascribeWith(true)))
+          case (x, Fix(:*:(left, right))) =>
+            HEnvT(
+              x,
+              :*:(
+                ascribeWith(false)(left),
+                ascribeWith(x)(right)
+              )
+            )
+          case (x, s) => HEnvT(x, s.unFix.hmap(ascribeWith(x)))
+        }
+    }
+
+  // This is needed in order to use `traverse` in `reads`.
+  implicit private val readsApplicative = new Applicative[JsResult] {
+
+    def point[A](a: => A): JsResult[A] = JsSuccess(a)
+
+    def ap[A, B](fa: => JsResult[A])(f: => JsResult[A => B]): JsResult[B] = (f, fa) match {
+      // Shamelessly copied from the play-json library
+      case (JsSuccess(f, _), JsSuccess(a, _)) => JsSuccess(f(a))
+      case (JsError(e1), JsError(e2))         => JsError(JsError.merge(e1, e2))
+      case (JsError(e), _)                    => JsError(e)
+      case (_, JsError(e))                    => JsError(e)
+
+    }
+  }
+
+  // This is needed to allow undefined optional fields to be treated as `None`.
+  final private def undefinedAsNull[A](field: String, r: Reads[A]): Reads[A] = Reads { json =>
+    ((json \ field) match {
+      case JsDefined(v) => r.reads(v)
+      case _            => r.reads(JsNull)
+    }).repath(JsPath \ field)
+  }
+
+  final private val labellingSeed =
+    new (FSchema[R.Prim, R.SumTermId, R.ProductTermId, ?] ~> LabelledSchema) {
+
+      def apply[A](fSchema: FSchema[R.Prim, R.SumTermId, R.ProductTermId, A]): LabelledSchema[A] =
+        (false, fSchema)
+    }
+
+  implicit final def reads(
+    implicit primNT: R.Prim ~> Reads,
+    branchLabel: R.SumTermId <~< String,
+    fieldLabel: R.ProductTermId <~< String
+  ): RInterpreter[Reads] =
+    new Interpreter[R.Prim, R.SumTermId, R.ProductTermId, Reads] {
+
+      val alg = new (HEnvT[Boolean, RSchema, Reads, ?] ~> Reads) {
+
+        def apply[A](schema: HEnvT[Boolean, RSchema, Reads, A]): Reads[A] = schema.fa match {
+          case One() =>
+            Reads {
+              case JsNull => JsSuccess(())
+              case _      => JsError(Seq(JsPath -> Seq(JsonValidationError("error.expected.null"))))
+            }
+          case :+:(left, right) =>
+            Reads(
+              json =>
+                left
+                  .reads(json)
+                  .fold(
+                    el =>
+                      right.reads(json).map(\/-.apply) match {
+                        case JsError(er) => JsError(JsError.merge(el, er))
+                        case x           => x
+                      },
+                    a => JsSuccess(-\/(a))
+                  )
+            )
```

Comment:
Would it be possible to directly choose the correct (`left` or `right`) decoder based on its `label`?

Reply:
Well, at that exact point it wouldn't, because a `:+:` node has no clue about what its children are "looking for" (and in the case of a `Union`, the label is handled by the `:+:`'s children).
But that's a good lead for future optimisation.